system/jlib/jtask.hpp

```diff
@@ -0,0 +1,174 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2022 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef JTASK_HPP
+#define JTASK_HPP
+
+#include <atomic>
+#include "jiface.hpp"
+#include "jthread.hpp"
+#include "jqueue.hpp"
+
+interface ITaskScheduler;
+
+class jlib_decl CTask : public CInterface
+{
+    friend class TaskQueue;
+    friend class DListOf<CTask>;
+
+public:
+    CTask(unsigned _numPred) : numPredecessors(_numPred) {}
+
+    //Return the next task to execute
+    virtual CTask * execute() = 0;
+
+    bool isReady() const { return numPredecessors == 0; }
+    void addPred()
+    {
+        numPredecessors.fetch_add(1);
+    }
+    // Return true if this is now available to execute.
+    bool notePredDone()
+    {
+        return numPredecessors.fetch_add(-1) == 1;
+    }
+
+    CTask * checkNextTask()
+    {
+        return nullptr;
+    }
+
+protected:
+    CTask * next = nullptr;
+    CTask * prev = nullptr;
+    std::atomic<unsigned> numPredecessors;
+};
+
+
+//---------------------------------------------------------------------------------------------------------------------
+
+interface ITaskScheduler : public IInterface
+{
+public:
+    virtual void enqueueOwnedTask(CTask * ownedTask) = 0;
+    virtual unsigned numProcessors() const = 0;
+};
+
+extern jlib_decl ITaskScheduler & queryTaskScheduler();
```

Comment:
I assume the expectation is that you pick a scheduler based on the characteristics of the task you are scheduling? A comment might help.